core/jvm/src/main/scala/zio/ZManagedPlatformSpecific.scala

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright 2017-2020 John A. De Goes and the ZIO Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package zio
+
+import java.io
+import java.io.IOException
+import java.net.{ URI, URL }
+import java.nio.file.Path
+
+import zio.blocking.{ Blocking, _ }
+
+private[zio] trait ZInputStream {
+  def readN(n: Int): ZIO[Blocking, IOException, Option[Chunk[Byte]]]
+  def skip(n: Long): ZIO[Blocking, IOException, Long]
+  def readAll: ZIO[Blocking, IOException, Option[Chunk[Byte]]]
+  def close(): ZIO[Blocking, Nothing, Unit]
+}
+
+private[zio] trait ZOutputStream {
+  def write(chunk: Chunk[Byte]): ZIO[Blocking, IOException, Unit]
+  def close(): ZIO[Blocking, Nothing, Unit]
+}
+
+/**
+ * A functional wrapper over a java.io.InputStream.
+ */
+private[zio] case class InputStream private (private val is: java.io.InputStream) extends ZInputStream {
+  def readN(n: Int): ZIO[Blocking, IOException, Option[Chunk[Byte]]] =
```

Comment:
With this signature, there's no way for the caller to infer if the input stream is exhausted. How about encoding this as `ZIO[Blocking, Option[IOException], Chunk[Byte]]`, where `Some(IOException)` is an actual error, `None` is an end-of-stream signal and a value is a chunk?

I would also suggest to make the non-blocking read optional (using a flag). If the flag is false, don't check `available` before reading and just let the underlying input stream manage that.

Reply:
Great suggestion. I think we should remove the available check alltogether if we use the encoding you proposed. And have each implementation decide how it wants to deal with readN on an exhausted stream. I think it makes things simpler.

 What do you think?